### PR TITLE
merkle-part-key-integration

### PR DESCRIPTION
**Pull request by algonathan**
## Summary
integrating the merklekeystore.Signer into participation key, and key registration

### CODE REVIEW

File: crypto/merklekeystore/keystore.go [hunk 1]
```diff
@@ -149,24 +166,83 @@ func (m *Signer) Sign(hashable crypto.Hashable, round uint64) (Signature, error)
...
+		mu:   deadlock.RWMutex{},
+	}
+
+	copy(signerCopy.EphemeralKeys.SignatureAlgorithms, m.EphemeralKeys.SignatureAlgorithms)
+	return &signerCopy
+}
+
+func (m *Signer) dropKeys(upTo int) {
+	if l := len(m.EphemeralKeys.SignatureAlgorithms); l < upTo {
+		upTo = l
```
[the remark this answers is not in the document]
Comment (algonathan) [reply crypto/merklekeystore/keystore.go:233]: when more than the max amount of keys is requested, the expected behaviour is to delete all. ( similar to onetimesig )

File: crypto/merklekeystore/keystore.go [hunk 2]
```diff
@@ -51,29 +56,41 @@ type (
 	Signature struct {
 		_struct              struct{} `codec:",omitempty,omitemptyarray"`
 		crypto.ByteSignature `codec:"bsig"`
-		Proof                `codec:"prf"`
-		VerifyingKey         crypto.VerifyingKey `codec:"vkey"`
+
+		Proof        `codec:"prf"`
+		VerifyingKey crypto.VerifyingKey `codec:"vkey"`
+		index        uint64
```
Comment (algoidan) [inline crypto/merklekeystore/keystore.go:62]: not sure you are using it

File: crypto/merklekeystore/keystore.go [hunk 3]
```diff
@@ -88,27 +105,43 @@ func (d *EphemeralKeys) Length() uint64 {
...
+	}
+
+	keys := make([]crypto.SignatureAlgorithm, numberOfKeys)
 	for i := range keys {
 		keys[i] = *crypto.NewSigner(sigAlgoType)
 	}
 	ephKeys := EphemeralKeys{
 		SignatureAlgorithms: keys,
-		firstRound:          firstValid,
+		TreeBase:            firstValid,
```
Comment (algoidan) [inline crypto/merklekeystore/keystore.go:143]: we should explicitly init ArrayBase value

File: crypto/merklekeystore/keystore.go [hunk 4]
```diff
@@ -149,32 +186,100 @@ func (m *Signer) Sign(hashable crypto.Hashable, round uint64) (Signature, error)
...
 	if round < firstValid {
-		return errOutOfBounds
+		return errReceivedRoundIsBeforeFirst
 	}
+
+	pos := roundToIndex(firstValid, round, interval)
 	ephkey := CommittablePublicKey{
 		VerifyingKey: sig.VerifyingKey,
-		Round:        round,
+		Round:        indexToRound(firstValid, interval, pos),
```
Comment (algoidan) [inline crypto/merklekeystore/keystore.go:276]: why not use "round"?

File: crypto/merklekeystore/keystore_test.go [hunk 5]
```diff
@@ -23,87 +23,169 @@ import (
...
+	a.Equal(0, len(signer.SignatureAlgorithms))
+
+	_, err = signer.Sign(genHashableForTest(), 2)
 	a.Error(err)
 
-	signer, err := New(0, 0, crypto.PlaceHolderType)
+	s, err := New(8, 21, 10, crypto.PlaceHolderType)
+	a.NoError(err)
+	a.Equal(len(s.SignatureAlgorithms), 2)
+
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:61]: add 
	s, err = New(8, 20, 10, crypto.PlaceHolderType)
	a.NoError(err)
	a.Equal(len(s.SignatureAlgorithms), 2)

File: crypto/merklekeystore/keystore_test.go [hunk 6]
```diff
@@ -23,87 +23,169 @@ import (
...
-	sig, err := signer.Sign(genHashableForTest(), 0)
+	a.Equal(0, len(signer.SignatureAlgorithms))
+
+	_, err = signer.Sign(h, 8)
+	a.Error(err)
+
+	_, err = signer.Sign(h, 9)
+	a.Error(err)
+
+	_, err = signer.Trim(10)
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:102]: add trim 9 too

File: crypto/merklekeystore/keystore_test.go [hunk 7]
```diff
@@ -23,87 +23,169 @@ import (
...
+	a.Error(err)
+	// take signer and make it empty, then try to sign with it:
+	signer, err = New(8, 9, 2, crypto.PlaceHolderType)
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:105]: no need for trim tests in here, you check that somewhere else.

File: crypto/merklekeystore/keystore_test.go [hunk 8]
```diff
@@ -23,87 +23,169 @@ import (
 
 	"github.com/algorand/go-algorand/crypto"
 	"github.com/algorand/go-algorand/protocol"
-
 	"github.com/stretchr/testify/require"
 )
 
 func TestSignerCreation(t *testing.T) {
 	a := require.New(t)
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:30]: add test for new signer with round == 0

File: crypto/merklekeystore/keystore_test.go [hunk 9]
```diff
@@ -155,13 +264,13 @@ func TestBadMerkleProofInSignature(t *testing.T) {
 
 	sig2 := sig
 	sig2.Proof = sig2.Proof[:len(sig2.Proof)-1]
-	a.Error(signer.GetVerifier().Verify(start, start, hashable, sig2))
+	a.Error(signer.GetVerifier().Verify(start, start, 1, hashable, sig2))
 
 	sig3 := sig2
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:269]: this might be an error, perhaps sig3 := sig

File: crypto/merklekeystore/keystore_test.go [hunk 10]
```diff
@@ -187,32 +296,126 @@ func TestAttemptToUseDifferentKey(t *testing.T) {
...
+	a.Equal(cpy.FirstValid, uint64(1))
+	a.Equal(cpy.ArrayBase, uint64(20))
+	a.Equal(len(cpy.SignatureAlgorithms), 80)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(len(signer.SignatureAlgorithms), 80)
+
+	_, err = signer.Trim(101)
+	a.Error(err)
+
+	signer, err = New(1, 10, 1, crypto.PlaceHolderType)
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:350]: delete

File: crypto/merklekeystore/keystore_test.go [hunk 11]
```diff
@@ -187,32 +296,126 @@ func TestAttemptToUseDifferentKey(t *testing.T) {
...
+	signer.Trim(22)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(7, len(signer.SignatureAlgorithms))
+
+	_, err = signer.Trim(23)
+	a.Error(err)
+
+	signer.Trim(99)
+	a.Equal(signer.FirstValid, uint64(1))
+	a.Equal(len(signer.SignatureAlgorithms), 0)
```
Comment (algoidan) [inline crypto/merklekeystore/keystore_test.go:371]: add another test: build signer and delete all keys.

File: data/account/account.go [hunk 12]
```diff
@@ -184,14 +185,10 @@ func RestoreParticipation(store db.Accessor) (acc PersistedParticipation, err er
...
+	acc.BlockProof = &merklekeystore.Signer{}
 	if err = protocol.Decode(rawBlockProof, acc.BlockProof); err != nil {
 		return PersistedParticipation{}, err
 	}
-	// rawBlockProofKey stored is invalid
-	if !acc.BlockProof.IsValid() {
-		return PersistedParticipation{}, fmt.Errorf("stored blockProof key is not valid")
-	}
-
+	// TODO: Should i check validity of the keys inside the Signer?
```
[the remark this answers is not in the document]
Comment (algonathan) [reply data/account/account.go:192]: im not sure what to do once we find out it wasn't stored in a valid state.
basically if someone created it using the code - it should be valid or empty.
if it isn't valid then someone did not use the code to store it. 
i don't think we should check that.

File: data/transactions/transaction.go [hunk 13]
```diff
@@ -569,11 +570,6 @@ func (tx Transaction) blockProofPKWellFormed(proto config.ConsensusParams) error
 		return errKeyRegEmptyBlockProofPK
 	}
```
[the remark this answers is not in the document]
Comment (algonathan) [reply data/transactions/transaction.go:572]: i think it is unneeded, the `isValid` is used to differ between empty and non empty.

**Pull request merged**